Mathlib/Logic/Equiv/Basic.lean

```diff
@@ -365,6 +366,26 @@ theorem sumAssoc_symm_apply_inr_inl {α β γ} (b) :
 theorem sumAssoc_symm_apply_inr_inr {α β γ} (c) : (sumAssoc α β γ).symm (inr (inr c)) = inr c :=
   rfl
 
+/-- Four-way commutativity of `sum`. The name matches `add_add_add_comm`. -/
+@[simps apply]
+def sumSumSumComm (α β γ δ) : (α ⊕ β) ⊕ γ ⊕ δ ≃ (α ⊕ γ) ⊕ β ⊕ δ where
+  toFun :=
+    (sumAssoc (α ⊕ γ) β δ) ∘ (Sum.map (sumAssoc α γ β).symm (@id δ))
+      ∘ (Sum.map (Sum.map (@id α) (sumComm β γ)) (@id δ))
+      ∘ (Sum.map (sumAssoc α β γ) (@id δ))
+      ∘ (sumAssoc (α ⊕ β) γ δ).symm
+  invFun :=
+    (sumAssoc (α ⊕ β) γ δ) ∘ (Sum.map (sumAssoc α β γ).symm (@id δ))
+      ∘ (Sum.map (Sum.map (@id α) (sumComm β γ).symm) (@id δ))
+      ∘ (Sum.map (sumAssoc α γ β) (@id δ))
+      ∘ (sumAssoc (α ⊕ γ) β δ).symm
+  left_inv x := by aesop
+  right_inv x := by aesop
```

Comment:
Remember, `∘` is not a reducible transparency, and `Function.comp` is not a `simp` lemma. You need to unfold it manually.
```suggestion
  left_inv x := by simp [Function.comp]
  right_inv x := by simp [Function.comp]
```
In fact, this makes me wonder slightly if your definitions of `toFun` and `invFun` should not be written in point-free style, but rather with a `match`. I'm not sure about the answer to this, maybe you can explore?

Reply:
I'm not sure if I understand your suggestion with a match. Can you say more about this?

Reply:
Yes, I'll get to it today.

Reply:
Sorry, I got distracted. This is what I meant, but I'm not saying it's actually better:
```lean
/-- Four-way commutativity of `sum`. The name matches `add_add_add_comm`. -/
@[simps apply]
def sumSumSumComm (α β γ δ) : (α ⊕ β) ⊕ γ ⊕ δ ≃ (α ⊕ γ) ⊕ β ⊕ δ where
  toFun := fun
    | inl (inl a) => inl (inl a)
    | inl (inr b) => inr (inl b)
    | inr (inl c) => inl (inr c)
    | inr (inr d) => inr (inr d)
  invFun := fun
    | inl (inl a) => inl (inl a)
    | inl (inr c) => inr (inl c)
    | inr (inl b) => inl (inr b)
    | inr (inr d) => inr (inr d)
  left_inv := by rintro ((a | b) | c | d); all_goals rfl
  right_inv := by rintro ((a | c) | b | d); all_goals rfl
```

Reply:
No worries; thanks for elaborating. I also like your definition. I'm not sure either which one is better. (Both would be nice to have.)
I'll merge this PR as-is; if you think this relation would be useful to have, just say so and I can file a follow-up PR to either change the definition (and add the equality with the current definition as lemmas) or vice versa).